akka-http2-support/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderDecompression.scala

```diff
@@ -42,13 +44,20 @@ private[http2] object HeaderDecompression extends GraphStage[FlowShape[FrameEven
       var headers = new VectorBuilder[(String, String)]
       object Receiver extends HeaderListener {
         def addHeader(name: Array[Byte], value: Array[Byte], sensitive: Boolean): Unit =
-          // TODO: optimization: use preallocated strings for well-known names, similar to
-          // what happens in HeaderParser
+          // TODO: optimization: use preallocated strings for well-known names, similar to what happens in HeaderParser
           headers += new String(name, UTF8) → new String(value, UTF8)
       }
-      decoder.decode(ByteStringInputStream(payload), Receiver)
-      decoder.endHeaderBlock() // TODO: do we have to check the result here?
-      push(eventsOut, ParsedHeadersFrame(streamId, endStream, headers.result()))
+      try {
+        decoder.decode(ByteStringInputStream(payload), Receiver)
+        decoder.endHeaderBlock() // TODO: do we have to check the result here?
+
+        push(eventsOut, ParsedHeadersFrame(streamId, endStream, headers.result()))
+      } catch {
+        case ex: IOException ⇒
+          // this is signalled by the decoder when it failed, we want to react to this by rendering a GO_AWAY frame
+          push(eventsOut, GoAwayFrame(streamId, ErrorCode.COMPRESSION_ERROR))
```

Comment:
In any case, we somehow need to distinguish a client-side GoAwayFrame from this one, right? In cases of errors, I'd rather use exceptions if we want to do the loop. Maybe let's try without the BidiStage as long as we can get away with it.